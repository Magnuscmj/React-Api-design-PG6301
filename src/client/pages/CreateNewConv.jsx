//This file has code from the lecturer and has been changed to fit this assignment
//https://github.com/kristiania-pg6301-2021/pg6301-react-and-express-lectures

import React, { useState } from "react";
import { InputField } from "../components/InputField";
import {Link, Redirect} from "react-router-dom";

export function CreateNewConv() {
    const [to, setTo] = useState("");
    const [text, setText] = useState("");
    const [time, setTime] = useState("");
    const [redirect, setRedirect] = useState(false);

    async function submit(e) {
        e.preventDefault();
        console.log("Submitting", { to, text, time });
        await fetch("/api/messages", {
            method: "POST",
            body: JSON.stringify({ to, text, time }),
            headers: {
                "Content-Type": "application/json",
            },
        });
        setRedirect(true);
    }
    if (redirect) {
        return <Redirect to='/messages' />;
    }

    return (

        <form onSubmit={submit}>
            <Link className="backBtn" to={"/Profile"}>Back</Link>
            <h1>Start new chat</h1>
            <InputField label={"To"} value={to} onChangeValue={setTo} />
            <InputField label={"Text"} value={text} onChangeValue={setText} />
            <InputField
                label={"Time"}
                value={time}
                onChangeValue={setTime}
                type="number"
            />
            <button>Submit</button>
        </form>
    );
}
