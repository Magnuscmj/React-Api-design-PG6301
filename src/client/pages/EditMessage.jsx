//This file has code from the lecturer and has been changed to fit this assignment
//https://github.com/kristiania-pg6301-2021/pg6301-react-and-express-lectures

import React, { useState } from "react";
import { useParams } from "react-router";
import { LoadingView } from "../components/LoadingView";
import { InputField } from "../components/InputField";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";
import {Link} from "react-router-dom";
import { Redirect } from 'react-router-dom';

function EditMessageForm({ message }) {
    const [to, setTo] = useState(message.to);
    const [text, setText] = useState(message.text);
    const [time, setTime] = useState(message.time);
    const [redirect, setRedirect] = useState(false);

    async function submit(e) {
        e.preventDefault();
        console.log("Submitting", { to, text, time });
        await fetch(`/api/messages/${message.id}`, {
            method: "PUT",
            body: JSON.stringify({ to, text, time }),
            headers: {
                "Content-Type": "application/json",
            },
        });
        setRedirect(true);
    }
    if (redirect) {
        return <Redirect to='/messages' />;
    }


return (
        <form onSubmit={submit}>
            <Link className="backBtn" to={"/Profile"}>Back</Link>
            <h1>Go to conversation ({to})</h1>
            <InputField label={"To"} value={to} onChangeValue={setTo} />
            <InputField label={"Text"} value={text} onChangeValue={setText} />
            <InputField
                label={"Time"}
                value={time}
                onChangeValue={setTime}
                type="number"
            />
            <button>Submit</button>
        </form>
    );
}

export function EditMessage({ messageApi }) {
    const { id } = useParams();

    const { data: message, loading, error, reload } = useLoading(
        async () => await messageApi.getMessage(id),
        [id]
    );

    if (error) {
        return <ErrorView error={error} reload={reload()} />;
    }

    if (loading || !message) {
        return <LoadingView />;
    }

    return <EditMessageForm message={message} />;
}



