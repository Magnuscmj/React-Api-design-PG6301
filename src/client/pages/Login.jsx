//This file has code from the lecturer and has been changed to fit this assignment
//https://github.com/kristiania-pg6301-2021/pg6301-react-and-express-lectures

import React, { useState } from "react";
import { postJson } from "../lib/http";
import { useHistory } from "react-router";
import { useSubmit } from "../components/UseSubmit";
import { InputField } from "../components/InputField";
import GoogleButton from "react-google-button";
import {Redirect} from "react-router-dom";

export function Login() {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
    const [redirect, setRedirect] = useState(false);
  const history = useHistory();

  const { handleSubmit: handleLogin, submitting, error } = useSubmit(
    async () => {
      await postJson("/api/loginHard",  { username, password });

        setRedirect(true);
        },
    () => history.push("/messages")
  );
    if (redirect) {
        return <Redirect to='/messages' />;
    }


  return (
    <div className="login-container">
      <h2 className="h2-sign-in">Please sign in</h2>
      <form className="form-login" onSubmit={handleLogin}>
        {submitting && <div>Please wait</div>}
        {error && <div>Error {error.toString()}</div>}

        <InputField
          className="username"
          id="floatingInput"
          type="username"
          autoFocus={true}
          value={username}
          onChangeValue={setUsername}
        />

        <div className="password">
          <InputField
            className="form-control"
            type="password"
            value={password}
            onChangeValue={setPassword}
          />
        </div>

        <button className="loginBtn" disabled={submitting}>
          Sign in
        </button>
      </form>

      <a href={"api/login"} target={"_self"}>
        <GoogleButton className="google-button" />
      </a>
    </div>
  );
}
