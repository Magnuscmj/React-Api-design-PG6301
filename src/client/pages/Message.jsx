//This file has code from the lecturer and has been changed to fit this assignment
//https://github.com/kristiania-pg6301-2021/pg6301-react-and-express-lectures

import React from "react";
import { LoadingView } from "../components/LoadingView";
import { Link } from "react-router-dom";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";


export function Message({ messageApi }) {
    const { data: messages, error, loading, reload } = useLoading(
        async () => await messageApi.listMessages()
    );

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !messages) {
        return <LoadingView />;
    }

    return (
        <div className="messagePage">
            <Link className="backBtn" to={"/profile"}>Back</Link>
            <h1>Chat</h1>
            {messages.map(({ id, to, text, time }) => (
                <li key={id}>
                    <Link to={`/messages/${id}/edit`}>({time})  {to}: {text}</Link>
                </li>
            ))}
        </div>
    );
}
