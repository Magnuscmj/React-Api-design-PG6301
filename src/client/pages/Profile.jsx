//This file has code from the lecturer and has been changed to fit this assignment
//https://github.com/kristiania-pg6301-2021/pg6301-react-and-express-lectures

import React from "react";
import { ErrorView } from "../components/ErrorView";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";
import { fetchJson } from "../lib/http";
import {Link} from "react-router-dom";

export function Profile() {
  const { loading, error, data } = useLoading(() => fetchJson("/api/profile"));


    function handleSubmitChatMessage(e) {
        e.preventDefault();

    }
  if (error) {
    return <ErrorView error={error} />;
  }
  if (loading || !data) {
    return <LoadingView />;
  }

  const { username} = data;



  return (
    <div>
      <h2>Your profile:</h2>
      <div>Username: {username}</div>
       <div className="messageList">
        <h1>Chat page</h1>
        <ul>
          <li>
            <Link to={"/messages"}>See messages</Link>
          </li>
          <li>
            <Link to={"/new"}>Start new chat</Link>
          </li>
        </ul>
       </div>
    </div>
  );
}
