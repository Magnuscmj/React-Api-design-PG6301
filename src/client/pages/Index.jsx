//This file has code from the lecturer and has been changed to fit this assignment
//https://github.com/kristiania-pg6301-2021/pg6301-react-and-express-lectures

import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";

import { Profile } from "./Profile";
import { Login } from "./Login";
import { Navbar } from "../components/Navbar";
import {Message} from "./Message";
import {CreateNewConv} from "./CreateNewConv";
import {EditMessage} from "./EditMessage";

function Application() {
    const messageApi = {
        listMessages: async () => {
            const res = await fetch("/api/messages");
            if (!res.ok) {
                throw new Error(
                    `Something went wrong loading ${res.url}: ${res.statusText}`
                );
            }
            return await res.json();
        },
        getMessage: async (id) => {
            const res = await fetch(`/api/messages/${id}`);
            if (!res.ok) {
                throw new Error(
                    `Something went wrong loading ${res.url}: ${res.statusText}`
                );
            }
            return await res.json();
        },
    };
  return (
    <BrowserRouter>
      <header>
        <Navbar />
      </header>
    <main>
      <Switch>
          <main className="form-signin">
          <Route exact path={"/messages"}>
              <Message messageApi={messageApi} />
          </Route>
          <Route path={"/new"}>
              <CreateNewConv />
          </Route>
          <Route path={"/messages/:id/edit"}>
              <EditMessage messageApi={messageApi} />
          </Route>
          <Route exact path={"/"}>
          </Route>

          <Route path="/profile" component={Profile} />
          <Route path="/login" component={Login} />
        </main>
        <Route>
          <h1>Not found</h1>
        </Route>
      </Switch>
    </main>
    </BrowserRouter>
  );
}

ReactDOM.render(<Application />, document.getElementById("app"));
