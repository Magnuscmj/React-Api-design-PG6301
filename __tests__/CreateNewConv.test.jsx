//This file has code from the lecturer and has been changed to fit this assignment
//https://github.com/kristiania-pg6301-2021/pg6301-react-and-express-lectures

import React from "react";
import * as ReactDOM from "react-dom";
import {MemoryRouter} from "react-router";
import {CreateNewConv} from "../src/client/pages/CreateNewConv";


describe("Create new messages", () => {
    it("Should render the new conversations", () => {
        const container = document.createElement("div");
        ReactDOM.render(
            <MemoryRouter>
                <CreateNewConv />,
            </MemoryRouter>,
            container
        );
        expect(container.innerHTML).toMatchSnapshot();
    })
})