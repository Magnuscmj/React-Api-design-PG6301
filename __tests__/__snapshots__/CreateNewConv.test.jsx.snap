// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create new messages Should render the new conversations 1`] = `"<form><a class=\\"backBtn\\" href=\\"/Profile\\">Back</a><h1>Start new chat</h1><div><label>To <input type=\\"text\\" placeholder=\\"text\\" value=\\"\\"></label></div><div><label>Text <input type=\\"text\\" placeholder=\\"text\\" value=\\"\\"></label></div><div><label>Time <input type=\\"number\\" placeholder=\\"number\\" value=\\"\\"></label></div><button>Submit</button></form>,"`;
